File: src/deadlock_build_sync/offline/discovery_data.py
"""Whole-match temporal partitions and inventories strictly before 20 minutes."""

from __future__ import annotations

from typing import TYPE_CHECKING

from .discovery_types import HeroLandmarkRow
from .sql_resources import load_sql

if TYPE_CHECKING:
    import duckdb

from collections import Counter, defaultdict
from dataclasses import dataclass, replace

import numpy as np

from deadlock_build_sync.build_support import SUPPORT
from deadlock_build_sync.mechanics import ItemGraph

from .inventory_reconstruction import reconstruct_final_inventory


@dataclass
class HeroDiscoveryData:
    hero: int
    items: tuple[int, ...]
    matrix: np.ndarray
    times: np.ndarray
    matches: np.ndarray
    folds: np.ndarray
    won: np.ndarray
    wealth: np.ndarray
    lead: np.ndarray
    badge: np.ndarray
    relative_wealth: np.ndarray
    enemies: np.ndarray
    actors: tuple[tuple[int, int], ...] = ()
    inventories: tuple[tuple[int, ...], ...] = ()

    def fold_mask(self, fold: str) -> np.ndarray:
        return self.folds == fold


def prepare_discovery_partitions(connection: duckdb.DuckDBPyConnection) -> None:
    if connection.execute(
        load_sql("discovery/count_split_boundaries.sql")
    ).fetchone() == (1,):
        connection.execute(load_sql("discovery/create_fixed_partitions.sql"))
        return
    connection.execute(load_sql("discovery/create_ranked_partitions.sql"))


def load_landmark_rows(
    connection: duckdb.DuckDBPyConnection, hero: int
) -> list[HeroLandmarkRow]:
    return connection.execute(
        load_sql("discovery/select_landmark_rows.sql"),
        {"hero": hero},
    ).fetchall()


def load_purchase_histories(
    connection: duckdb.DuckDBPyConnection, hero: int
) -> dict[tuple[int, int], list[tuple[int, int, int]]]:
    rows = connection.execute(
        load_sql("discovery/select_purchase_histories.sql"),
        {"hero": hero},
    ).fetchall()
    histories: dict[tuple[int, int], list[tuple[int, int, int]]] = defaultdict(list)
    for match, slot, item, bought, sold in rows:
        histories[int(match), int(slot)].append((
            int(item),
            int(bought),
            int(sold) if sold and sold < 1200 else 0,
        ))
    return dict(histories)


def build_hero_discovery_data(
    hero: int,
    rows: list[HeroLandmarkRow],
    histories: dict[tuple[int, int], list[tuple[int, int, int]]],
    graph: ItemGraph,
) -> HeroDiscoveryData:
    actors = tuple((int(row[0]), int(row[1])) for row in rows)
    if len(set(actors)) != len(actors) or len({row[0] for row in actors}) != len(
        actors
    ):
        raise ValueError("Discovery contains duplicate hero appearances in a match")
    inventories = tuple(
        reconstruct_final_inventory(histories.get(actor, []), graph.components)
        for actor in actors
    )
    support: Counter[int] = Counter()
    for row, owned in zip(rows, inventories, strict=True):
        if row[2] == "discovery":
            support.update(set(owned))
    minimum = max(
        SUPPORT.core_owners,
        int(np.ceil(sum(row[2] == "discovery" for row in rows) * 0.01)),
    )
    items = tuple(
        sorted(
            item
            for item, count in support.items()
            if graph.require(item).cost >= 1600 and count >= minimum
        )
    )
    times = np.full((len(rows), len(items)), -1, dtype=np.int32)
    index = {item: column for column, item in enumerate(items)}
    for offset, (actor, owned) in enumerate(zip(actors, inventories, strict=True)):
        for item, bought, _ in histories.get(actor, []):
            if item in index and item in owned:
                times[offset, index[item]] = max(times[offset, index[item]], bought)
    return replace(
        _build_hero_arrays(hero, items, rows, times),
        actors=actors,
        inventories=inventories,
    )


def _build_hero_arrays(
    hero: int,
    items: tuple[int, ...],
    rows: list[HeroLandmarkRow],
    times: np.ndarray,
) -> HeroDiscoveryData:
    return HeroDiscoveryData(
        hero,
        items,
        times >= 0,
        times,
        np.asarray([row[0] for row in rows], dtype=np.int64),
        np.asarray([row[2] for row in rows], dtype="U12"),
        np.asarray([row[3] for row in rows], dtype=bool),
        np.asarray([row[4] for row in rows], dtype=float),
        np.asarray([row[5] for row in rows], dtype=float),
        np.asarray([row[6] for row in rows], dtype=int),
        np.asarray([row[7] for row in rows], dtype=float),
        np.asarray(
            [
                row[8] if row[8] is not None and len(row[8]) == 6 else [0] * 6
                for row in rows
            ],
            dtype=np.int64,
        ).reshape(-1, 6),
    )


def load_hero_discovery_data(
    connection: duckdb.DuckDBPyConnection,
    hero: int,
    graph: ItemGraph,
    minimum: int = 11,
    maximum: int = 116,
) -> HeroDiscoveryData:
    count = connection.execute(
        load_sql("discovery/count_hero_appearances.sql"), {"hero": hero}
    ).fetchone()
    if count is None or count[0] == 0:
        raise ValueError(f"Hero {hero} has no source data; run refresh-evidence again")
    return build_hero_discovery_data(
        hero,
        [
            row
            for row in load_landmark_rows(connection, hero)
            if minimum <= row[6] <= maximum
        ],
        load_purchase_histories(connection, hero),
        graph,
    )

File: crates/deadlock-analysis/src/discovery_data.rs
use std::collections::{BTreeMap, BTreeSet};

use deadlock_data::{Error, RankRange, Result};
use deadlock_input::ItemGraph;
use serde::Deserialize;

use crate::database::{AnalysisDatabase, Parameters};
use crate::inventory_history::{Actor, Purchase, inventory_before};
use crate::sql_resources::load_sql;

#[derive(Clone, Debug, Deserialize)]
pub struct Landmark {
    pub match_id: u64,
    pub player_slot: u64,
    #[serde(rename = "partition")]
    pub fold: String,
    pub won: bool,
    pub observed_wealth: Option<f64>,
    pub team_lead: Option<f64>,
    pub average_badge: u16,
    pub relative_wealth: Option<f64>,
}

#[derive(Clone, Debug)]
pub struct DiscoveryData {
    pub hero: u64,
    pub items: Vec<u64>,
    pub times: Vec<Vec<i64>>,
    pub rows: Vec<Landmark>,
    pub inventories: Vec<Vec<u64>>,
}

impl DiscoveryData {
    pub fn fold_rows<'a>(&'a self, fold: &'a str) -> impl Iterator<Item = usize> + 'a {
        self.rows
            .iter()
            .enumerate()
            .filter(move |(_, row)| row.fold == fold)
            .map(|(index, _)| index)
    }

    pub fn columns(&self, items: &[u64]) -> Result<Vec<usize>> {
        items
            .iter()
            .map(|item| {
                self.items
                    .binary_search(item)
                    .map_err(|_| Error::new("Core contains an item outside the discovery matrix"))
            })
            .collect()
    }

    pub fn owns(&self, row: usize, columns: &[usize]) -> bool {
        columns.iter().all(|column| self.times[row][*column] >= 0)
    }

    pub fn owners(&self, items: &[u64], folds: &[&str]) -> Result<BTreeSet<Actor>> {
        let columns = self.columns(items)?;
        Ok(self
            .rows
            .iter()
            .enumerate()
            .filter(|(index, row)| {
                folds.contains(&row.fold.as_str()) && self.owns(*index, &columns)
            })
            .map(|(_, row)| (row.match_id, row.player_slot))
            .collect())
    }
}

pub fn prepare_partitions(database: &AnalysisDatabase) -> Result<()> {
    let fixed = database.count(
        load_sql("discovery/count_split_boundaries.sql")?,
        &Parameters::new(),
    )? == 1;
    let query = if fixed {
        "discovery/create_fixed_partitions.sql"
    } else {
        "discovery/create_ranked_partitions.sql"
    };
    database.execute(load_sql(query)?, &Parameters::new())
}

pub fn load_discovery_data(
    database: &AnalysisDatabase,
    hero: u64,
    graph: &ItemGraph,
    ranks: RankRange,
) -> Result<DiscoveryData> {
    let parameters = Parameters::from([("hero".into(), duckdb::types::Value::UBigInt(hero))]);
    if database.count(
        load_sql("discovery/count_hero_appearances.sql")?,
        &parameters,
    )? == 0
    {
        return Err(Error::new(format!(
            "Hero {hero} has no source data; run refresh-evidence"
        )));
    }
    let mut parameters = parameters;
    parameters.insert("ownership_before_seconds".into(), 1200_i64.into());
    let rows = database
        .query_rows::<Landmark>(load_sql("discovery/select_landmark_rows.sql")?, &parameters)?
        .into_iter()
        .filter(|row| (ranks.minimum.badge()..=ranks.maximum.badge()).contains(&row.average_badge))
        .collect::<Vec<_>>();
    let mut histories = BTreeMap::<Actor, Vec<Purchase>>::new();
    database.visit_rows(
        load_sql("discovery/select_purchase_histories.sql")?,
        &parameters,
        |purchase: Purchase| {
            histories
                .entry((purchase.match_id, purchase.player_slot))
                .or_default()
                .push(purchase);
            Ok(())
        },
    )?;
    build_discovery_data(hero, rows, &histories, graph)
}

fn build_discovery_data(
    hero: u64,
    rows: Vec<Landmark>,
    histories: &BTreeMap<Actor, Vec<Purchase>>,
    graph: &ItemGraph,
) -> Result<DiscoveryData> {
    if rows
        .iter()
        .map(|row| row.match_id)
        .collect::<BTreeSet<_>>()
        .len()
        != rows.len()
    {
        return Err(Error::new(
            "Discovery contains duplicate hero appearances in a match",
        ));
    }
    let inventories = rows
        .iter()
        .map(|row| {
            inventory_before(
                histories
                    .get(&(row.match_id, row.player_slot))
                    .map_or(&[], Vec::as_slice),
                graph,
                1200,
            )
        })
        .collect::<Result<Vec<_>>>()?;
    let mut support = BTreeMap::<u64, usize>::new();
    for (row, owned) in rows
        .iter()
        .zip(&inventories)
        .filter(|(row, _)| row.fold == "discovery")
    {
        let _ = row;
        let mut distinct = owned.clone();
        distinct.sort_unstable();
        distinct.dedup();
        for item in distinct {
            *support.entry(item).or_default() += 1;
        }
    }
    let minimum = rows
        .iter()
        .filter(|row| row.fold == "discovery")
        .count()
        .div_ceil(100)
        .max(100);
    let items = support
        .into_iter()
        .filter(|(_, count)| *count >= minimum)
        .map(|(item, _)| Ok((item, graph.require(item)?.cost)))
        .collect::<Result<Vec<_>>>()?
        .into_iter()
        .filter(|(_, cost)| *cost >= 1600)
        .map(|(item, _)| item)
        .collect::<Vec<_>>();
    let mut times = vec![vec![-1; items.len()]; rows.len()];
    for (index, row) in rows.iter().enumerate() {
        for purchase in histories
            .get(&(row.match_id, row.player_slot))
            .map_or(&[][..], Vec::as_slice)
        {
            if let Ok(column) = items.binary_search(&purchase.item_id)
                && inventories[index].contains(&purchase.item_id)
            {
                times[index][column] = times[index][column].max(purchase.buy_time);
            }
        }
    }
    Ok(DiscoveryData {
        hero,
        items,
        times,
        rows,
        inventories,
    })
}
